File: src/main/java/ru/falseteam/neural2048/nn/NeuralNetworkManager.java
package ru.falseteam.neural2048.nn;

import ru.falseteam.neural2048.gnn.GeneticNeuralNetwork;
import ru.falseteam.neural2048.nn.ThresholdFunction;

import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBException;
import javax.xml.bind.Marshaller;
import javax.xml.bind.Unmarshaller;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Random;

/**
 * Сохранение, загрузка и генерация нейронных сетей
 *
 * @author Vladislav Sumin
 * @version 1.0
 */
public class NeuralNetworkManager {
    private static final Random rnd = new Random();

    /**
     * Создает нейронную сеть по заданным параметрам.
     * Каждый нейронн связывается со всемы нейронами следующего слоя
     *
     * @param functions - список функций активации (функции из этого списка выбираются случайно для каждого нейрона)
     * @param config    - конфигурация нейронов
     */
    public static NeuralNetwork createNeuralNetwork(ThresholdFunction[] functions, int[] config) {
        int countOfNeuron = 0;
        for (int i : config) {
            countOfNeuron += i;
        }
        NeuralNetwork nn = new GeneticNeuralNetwork(countOfNeuron);

        for (int i = 0; i < countOfNeuron; i++) {
            ThresholdFunction f = functions[rnd.nextInt(functions.length)];
            nn.setNeuronFunction(i, f, f.getDefaultParams());
        }

        int offset = config[0];
        for (int i = 1; i < config.length; i++) {
            for (int j = 0; j < config[i - 1]; j++) {
                for (int k = 0; k < config[i]; k++) {
                    nn.addLink(offset - 1 - j, offset + k, getRandomWeight());
                }
            }
            offset += config[i];
        }
        return nn;
    }

    public static void save(NeuralNetwork nn, OutputStream stream) throws JAXBException {
        JAXBContext context = JAXBContext.newInstance(NeuralNetwork.class);
        Marshaller marshaller = context.createMarshaller();
        marshaller.marshal(nn, stream);
    }

    public static NeuralNetwork load(InputStream stream)
            throws JAXBException {
        JAXBContext context = JAXBContext.newInstance(NeuralNetwork.class);
        Unmarshaller marshaller = context.createUnmarshaller();
        NeuralNetwork unmarshal = (NeuralNetwork) marshaller.unmarshal(stream);
        return unmarshal;
    }

    /**
     * Генерирует случайное значение веса для нейрона
     *
     * @return - вес
     */
    private static int getRandomWeight() {
        final int maxWeightNum = 10;
        return rnd.nextInt(maxWeightNum) - rnd.nextInt(maxWeightNum);
    }
}

File: src/main/java/ru/falseteam/neural2048/nn/NeuralNetworkManager.kt
package ru.falseteam.neural2048.nn

import javax.xml.bind.JAXBContext
import javax.xml.bind.JAXBException
import java.io.InputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.io.OutputStream
import java.util.Random

/**
 * Сохранение, загрузка и генерация нейронных сетей
 *
 * @author Vladislav Sumin
 * @version 1.0
 */
object NeuralNetworkManager {
    private val rnd = Random()

    /**
     * Создает нейронную сеть по заданным параметрам.
     * Каждый нейрон связывается со всеми нейронами следующего слоя
     *
     * @param functions - список функций активации (функции из этого списка выбираются случайно для каждого нейрона)
     * @param config    - конфигурация нейронов
     */
    fun createNeuralNetwork(functions: Array<ThresholdFunction>, config: IntArray): NeuralNetwork {
        val countOfNeuron = config.sum()
        val nn = NeuralNetwork(countOfNeuron)

        for (i in 0 until countOfNeuron) {
            val f = functions[rnd.nextInt(functions.size)]
            nn.setNeuronFunction(i, f, f.defaultParams)
        }

        var offset = config[0]
        for (i in 1 until config.size) {
            for (j in 0 until config[i - 1]) {
                for (k in 0 until config[i]) {
                    nn.addLink(offset - 1 - j, offset + k, getRandomWeight().toDouble())
                }
            }
            offset += config[i]
        }
        return nn
    }

    @Throws(JAXBException::class)
    fun saveAsXML(nn: NeuralNetwork, stream: OutputStream) {
        val context = JAXBContext.newInstance(NeuralNetwork::class.java)
        val marshaller = context.createMarshaller()
        marshaller.marshal(nn, stream)
    }

    @Throws(JAXBException::class)
    fun loadFromXML(stream: InputStream): NeuralNetwork {
        val context = JAXBContext.newInstance(NeuralNetwork::class.java)
        val marshaller = context.createUnmarshaller()
        return marshaller.unmarshal(stream) as NeuralNetwork
    }

    fun saveAsBinary(nn: NeuralNetwork, stream: OutputStream) {
        val out = ObjectOutputStream(stream)
        out.writeObject(nn)
        out.close()
    }

    fun loadFromBinary(stream: InputStream): NeuralNetwork {
        val input = ObjectInputStream(stream)
        val nn = input.readObject() as NeuralNetwork
        input.close()
        return nn
    }

    /**
     * Генерирует случайное значение веса для нейрона
     *
     * @return - вес
     */
    private fun getRandomWeight(): Int {
        val maxWeightNum = 10
        return rnd.nextInt(maxWeightNum) - rnd.nextInt(maxWeightNum)
    }
}
